server/neptune/gateway/event/push_event_handler.go

```diff
@@ -149,18 +160,70 @@ func (p *PushHandler) generatePlanSteps(repoID string) []workflows.Step {
 	return workflowSteps
 }
 
-func (p *PushHandler) generateApplySteps(repoID string) []workflows.Step {
-	var workflowSteps []workflows.Step
-	projectConfig := p.GlobalCfg.DefaultProjCfg(p.Logger, repoID, "path", defaultWorkspace)
-	steps := projectConfig.Workflow.Apply.Steps
-	for _, step := range steps {
-		workflowSteps = append(workflowSteps, workflows.Step{
-			StepName:    step.StepName,
-			ExtraArgs:   step.ExtraArgs,
-			RunCommand:  step.RunCommand,
-			EnvVarName:  step.EnvVarName,
-			EnvVarValue: step.EnvVarValue,
-		})
+func (p *PushHandler) buildRoots(event Push, ctx context.Context) ([]*valid.MergedProjectCfg, error) {
+	err := p.PreWorkflowHooksCommandRunner.RunPreHooksWithSha(ctx, p.Logger, event.Repo, event.Sha)
+	if err != nil {
+		p.Logger.Error(fmt.Sprintf("Error running pre-workflow hooks %s. Proceeding with root building.", err))
 	}
-	return workflowSteps
+
+	modifiedFiles, err := p.VCSClient.GetModifiedFilesFromCommit(event.Repo, event.Sha)
+	if err != nil {
+		return nil, err
+	}
+
+	unlockFn, err := p.WorkingDirLocker.TryLockOnSha(event.Repo.FullName, event.Sha, DefaultWorkspace)
+	if err != nil {
+		return nil, errors.Wrap(err, "locking working dir")
+	}
+	defer unlockFn()
+	repoDir, err := p.WorkingDir.CloneFromSha(p.Logger, event.Repo, event.Sha, DefaultWorkspace)
+	if err != nil {
+		return nil, err
+	}
+	deleteFn := func() {
+		if err := p.WorkingDir.DeleteSha(event.Repo, event.Sha); err != nil {
+			p.Logger.Error("failed deleting cloned repo", map[string]interface{}{
+				"err": err,
+			})
+		}
+	}
+	defer deleteFn()
+
+	// Parse config file if it exists.
+	hasRepoCfg, err := p.ParserValidator.HasRepoCfg(repoDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "looking for %s file in %q", config.AtlantisYAMLFilename, repoDir)
+	}
+
+	var mergedProjectCfgs []*valid.MergedProjectCfg
+	if hasRepoCfg {
+		// If there's a repo cfg then we'll use it to figure out which projects
+		// should be planed.
+		repoCfg, err := p.ParserValidator.ParseRepoCfg(repoDir, p.GlobalCfg, event.Repo.ID())
+		if err != nil {
+			return nil, errors.Wrapf(err, "parsing %s", config.AtlantisYAMLFilename)
+		}
+		matchingProjects, err := p.ProjectFinder.DetermineProjectsViaConfig(p.Logger, modifiedFiles, repoCfg, repoDir)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, mp := range matchingProjects {
+			mergedProjectCfg := p.GlobalCfg.MergeProjectCfg(p.Logger, event.Repo.ID(), mp, repoCfg)
+			mergedProjectCfgs = append(mergedProjectCfgs, &mergedProjectCfg)
+		}
+	} else {
```

Comment:
I'm of the opinion that we should just fail if we don't have a repo config.  Especially because we have entire control on how we write this.  This will just save issues we have with nil parameters that are guaranteed in our repo configs for example and tightens our assumptions.